pkg/logql/log/filter.go

```diff
@@ -6,20 +6,41 @@ import (
 	"unicode"
 	"unicode/utf8"
 
-	"github.com/grafana/loki/pkg/util"
-
 	"github.com/grafana/regexp"
 	"github.com/grafana/regexp/syntax"
 
+	"github.com/grafana/loki/pkg/util"
 	"github.com/prometheus/prometheus/model/labels"
 )
 
+type Checker interface {
+	Test(line []byte, caseInsensitive bool, equal bool) bool
+	TestRegex(reg *regexp.Regexp) bool
+}
+
+type Matcher interface {
+	Matches(test Checker) bool
+}
+
 // Filterer is a interface to filter log lines.
 type Filterer interface {
 	Filter(line []byte) bool
 	ToStage() Stage
 }
 
+type MatcherFilterer interface {
+	Matcher
+	Filterer
+}
+
+type matcherFiltererWrapper struct {
+	Filterer
+}
+
+func (m matcherFiltererWrapper) Matches(_ Checker) bool {
+	panic("unimplemented")
+}
```

Comment:
nit: instead of an "unidirectional" wrapper, we could make a "bidirectional" wrapper, for both `Filterer` and `Matcher`.

That could then also be used in the bloom package, right?

```go

type wrapper struct {
	Filterer
	Matcher
}

func (w wrapper) IsMatcher() bool {
	return a.Matcher != nil
}

func (w wrapper) IsFilterer() bool {
	return a.Filterer != nil
}

func WrapMatcher(f Filterer) MatcherFilterer {
	return adapter{Filterer: f}
}

func WrapFilterer(m Matcher) MatcherFilterer {
	return adapter{Matcher: m}
}
```

Reply:
Done! Thanks for the tip :)